features/recognize.js

```diff
@@ -18,68 +18,121 @@ module.exports = function (controller) {
   controller.hears(
     recognizeEmoji,
     ["direct_message", "direct_mention", "mention", "message"],
-    async (bot, message) => {
-      const userInfo = await userDetails(bot, message, message.user);
-      if (!userInfo) {
-        // TODO Error handling
-        return;
-      }
-
-      const errors = await checkForRecognitionErrors(message, userInfo);
-      if (errors) {
-        await bot.replyEphemeral(
-          message,
-          [
-            `Sending ${recognizeEmoji} failed with the following error(s):`,
-            errors,
-          ].join("\n")
-        );
-        return;
-      }
-
-      await sendRecognition(message, userInfo);
-      const gratitudeRemaining = await balance.dailyGratitudeRemaining(
-        userInfo.giver.id,
-        userInfo.giver.tz
-      );
-
-      await Promise.all([
-        sendNotificationToReceivers(bot, message, userInfo),
-        bot.replyEphemeral(
-          message,
-          `Your ${recognizeEmoji} has been sent. You have ${gratitudeRemaining} left to give today.`
-        ),
-      ]);
-    }
+    respondToRecognitionMessage
   );
 
-  controller.on("reaction_added", testFunction);
+  controller.on("reaction_added", respondToRecognitionReaction);
 };
 
-async function testFunction(bot, message) {
+async function respondToRecognitionMessage(bot, message) {
+  winston.info(`Heard reference to ${recognizeEmoji}`, {
+    callingUser: message.user,
+    slackMessage: message.text,
+  });
+
+  let userInfo;
+  try {
+    userInfo = await userDetails(bot, message.text, message.user);
+  } catch (err) {
+    winston.error("Slack API returned error from users.info", {
+      callingUser: message.user,
+      slackMessage: message.text,
+      APIResponse: err,
+    });
+    await bot.replyEphemeral(
+      message,
+      `Something went wrong while sending recognition. When retreiving user information from Slack, the API responded with the following error: \n ${JSON.stringify(
```

Comment:
Is there a chance that `userDetails` (or a function that it calls) throws an `Error` object? If it can, then I think the result of `JSON.stringify` is an empty object:

```
> a = new Error('foo')
Error: foo
    at REPL56:1:5
    at Script.runInThisContext (vm.js:132:18)
    at REPLServer.defaultEval (repl.js:479:29)
    at bound (domain.js:430:14)
    at REPLServer.runBound [as eval] (domain.js:443:12)
    at REPLServer.onLine (repl.js:811:10)
    at REPLServer.emit (events.js:326:22)
    at REPLServer.EventEmitter.emit (domain.js:486:12)
    at REPLServer.Interface._onLine (readline.js:337:10)
    at REPLServer.Interface._line (readline.js:666:8)
> JSON.stringify(a)
'{}'
> a.stack
'Error: foo\n' +
  '    at REPL56:1:5\n' +
  '    at Script.runInThisContext (vm.js:132:18)\n' +
  '    at REPLServer.defaultEval (repl.js:479:29)\n' +
  '    at bound (domain.js:430:14)\n' +
  '    at REPLServer.runBound [as eval] (domain.js:443:12)\n' +
  '    at REPLServer.onLine (repl.js:811:10)\n' +
  '    at REPLServer.emit (events.js:326:22)\n' +
  '    at REPLServer.EventEmitter.emit (domain.js:486:12)\n' +
  '    at REPLServer.Interface._onLine (readline.js:337:10)\n' +
  '    at REPLServer.Interface._line (readline.js:666:8)'
```

Reply:
As far as I know it shouldn't in any cases, it should just be throwing some JSON returned by the Slack API. However, this error handling was a pretty quick and lazy setup, so I'll try and take a second look to see if there's something better to be done here.